File: backend/app/schemas/task.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict


class TaskStatus(str, Enum):
    TODO = "TODO"
    IN_PROGRESS = "IN_PROGRESS"
    IN_REVIEW = "IN_REVIEW"
    DONE = "DONE"
    BLOCKED = "BLOCKED"
    ON_HOLD = "ON_HOLD"
    CANCELLED = "CANCELLED"


class TaskPriority(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    project_id: int
    assigned_to: Optional[int] = None
    status: TaskStatus = TaskStatus.TODO
    priority: TaskPriority = TaskPriority.MEDIUM
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    estimated_hours: Optional[Decimal] = None
    actual_hours: Optional[Decimal] = None


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    assigned_to: Optional[int] = None
    status: Optional[TaskStatus] = None
    priority: Optional[TaskPriority] = None
    start_date: Optional[date] = None
    due_date: Optional[date] = None
    estimated_hours: Optional[Decimal] = None
    actual_hours: Optional[Decimal] = None
    completed_at: Optional[datetime] = None


class TaskResponse(TaskBase):
    id: int
    company_id: int
    created_by: int
    updated_by: Optional[int] = None
    completed_at: Optional[datetime] = None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/pages/dashboard/tasks/types/task.ts
export interface Task {
  id: number;

  company_id: number;

  project_id: number;

  assigned_to: number | null;

  created_by: number;

  updated_by: number | null;

  title: string;

  description: string | null;

  status: string;

  priority: string;

  start_date: string;

  due_date: string | null;

  estimated_hours: number | null;

  actual_hours: number | null;

  created_at: string;

  updated_at: string | null;
}

export interface TaskCreate {
  project_id: number;

  assigned_to?: number | null;

  title: string;

  description?: string;

  status?: string;

  priority?: string;

  start_date: string;

  due_date?: string | null;

  estimated_hours?: number | null;

  actual_hours?: number | null;
}

export interface TaskUpdate {
  project_id?: number;

  assigned_to?: number | null;

  title?: string;

  description?: string;

  status?: string;

  priority?: string;

  start_date?: string;

  due_date?: string | null;

  estimated_hours?: number | null;

  actual_hours?: number | null;
}